testing/trino-product-tests/src/main/java/io/trino/tests/product/utils/DeltaQueryExecutor.java

```diff
@@ -0,0 +1,143 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.tests.product.utils;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.collect.ImmutableList;
+import dev.failsafe.Failsafe;
+import dev.failsafe.RetryPolicy;
+import io.airlift.log.Logger;
+import io.airlift.resolver.ArtifactResolver;
+import io.airlift.resolver.DefaultArtifact;
+import io.trino.tempto.context.TestContext;
+import io.trino.tempto.query.JdbcConnectionsPool;
+import io.trino.tempto.query.JdbcConnectivityParamsState;
+import io.trino.tempto.query.JdbcQueryExecutor;
+import org.sonatype.aether.artifact.Artifact;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.sql.Connection;
+import java.sql.Driver;
+import java.sql.SQLException;
+import java.time.Duration;
+import java.util.List;
+import java.util.Properties;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.airlift.resolver.ArtifactResolver.MAVEN_CENTRAL_URI;
+import static io.airlift.resolver.ArtifactResolver.USER_LOCAL_REPO;
+import static io.trino.collect.cache.CacheUtils.uncheckedCacheGet;
+import static io.trino.collect.cache.SafeCaches.buildNonEvictableCache;
+import static java.lang.ClassLoader.getPlatformClassLoader;
+
+public class DeltaQueryExecutor
+        extends JdbcQueryExecutor
+{
+    private static final Logger log = Logger.get(DeltaQueryExecutor.class);
+
+    private static final Cache<String, Driver> DRIVERS = buildNonEvictableCache(CacheBuilder.newBuilder().maximumSize(2));
+    private static final DeltaConnectionsPool CONNECTIONS_POOL = new DeltaConnectionsPool();
+
+    private static final RetryPolicy<Driver> loadDatabaseDriverRetryPolicy = RetryPolicy.<Driver>builder()
+            .withMaxRetries(30)
+            .withDelay(Duration.ofSeconds(10))
+            .onRetry(event -> log.warn(event.getLastException(), "Download failed on attempt %d, will retry.", event.getAttemptCount()))
+            .build();
+
+    public DeltaQueryExecutor(TestContext testContext)
+    {
+        super(testContext.getDependency(JdbcConnectivityParamsState.class, "delta"), CONNECTIONS_POOL, testContext);
+    }
+
+    private static class DeltaConnectionsPool
+            extends JdbcConnectionsPool
+    {
+        private Connection connection;
+
+        @Override
+        public Connection createConnection(JdbcConnectivityParamsState jdbcParamsState)
+                throws SQLException
+        {
+            if (connection == null || connection.isClosed()) {
```

Comment:
i think was i missed in https://github.com/trinodb/tempto/pull/116 is that we want to initialize pooling data source with the driver, perhaps using `org.apache.commons.dbcp2.BasicDataSource#setDriver`
https://github.com/trinodb/tempto/pull/117 could help

if we want equivalent of that, we 

- should keep connections in a map, key'd by `JdbcConnectivityParamsState`
- make sure they are eventually closed
- make sure we don't give one connection to two different callers (leases)

However, we maybe don't need any of this. 

The simplest approach would be to implement `QueryExecutor` interface directly, and then we could manage permanent connection more simply. The query executor is instantiated for given `JdbcConnectivityParamsState`. However, implementing `QueryExecutor` interface directly isn't practical. `JdbcQueryExecutor` has a lot of logic around query execution that we want to inherit. The simplest way to extend that class would be to plug into `JdbcQueryExecutor.openConnection` and bypass `JdbcConnectionsPool`.


```java
public class DeltaQueryExecutor
        extends JdbcQueryExecutor
{
    private static final Logger log = Logger.get(DeltaQueryExecutor.class);

    @GuardedBy("DRIVERS")
    private static final Map<String, Driver> DRIVERS = new HashMap<>();

    private static final RetryPolicy<Driver> loadDatabaseDriverRetryPolicy = RetryPolicy.<Driver>builder()
            .withMaxRetries(30)
            .withDelay(Duration.ofSeconds(10))
            .onRetry(event -> log.warn(event.getLastException(), "Download failed on attempt %d, will retry.", event.getAttemptCount()))
            .build();

    private final JdbcConnectivityParamsState jdbcParamsState;
    private final Field connectionField;

    public DeltaQueryExecutor(TestContext testContext)
    {
        super(testContext.getDependency(JdbcConnectivityParamsState.class, "delta"), mockConnectionsPool(), testContext);
        jdbcParamsState = testContext.getDependency(JdbcConnectivityParamsState.class, "delta");
        try {
            connectionField = getClass().getSuperclass().getDeclaredField("connection");
            connectionField.setAccessible(true);
        }
        catch (ReflectiveOperationException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void openConnection()
    {
        closeConnection();

        Driver driver;
        synchronized (DRIVERS) {
            driver = DRIVERS.computeIfAbsent(jdbcParamsState.driverClass, className -> Failsafe.with(loadDatabaseDriverRetryPolicy)
                    .get(() -> loadDatabaseDriver(className)));
        }

        Properties properties = new Properties();
        properties.put("user", jdbcParamsState.user);
        properties.put("password", jdbcParamsState.password);
        try {
            Connection connection = driver.connect(jdbcParamsState.url, properties);
            requireNonNull(connection, "connection is null");
            connectionField.set(this, connection);
        }
        catch (SQLException | ReflectiveOperationException e) {
            throw new RuntimeException(e);
        }
    }

    private static Driver loadDatabaseDriver(String driverClassName)
    {
        // TODO Add support for maven coordinate in tempto
        ArtifactResolver resolver = new ArtifactResolver(USER_LOCAL_REPO, ImmutableList.of(MAVEN_CENTRAL_URI));
        List<URL> classPath = resolver.resolveArtifacts(getArtifact(driverClassName)).stream()
                .map(artifact -> {
                    try {
                        return artifact.getFile().toURI().toURL();
                    }
                    catch (MalformedURLException e) {
                        throw new RuntimeException(e);
                    }
                })
                .collect(toImmutableList());
        checkArgument(!classPath.isEmpty(), "classPath must not be empty");

        @SuppressWarnings("resource")
        URLClassLoader classLoader = new URLClassLoader(classPath.toArray(URL[]::new), getPlatformClassLoader());
        Class<? extends Driver> driverClass;
        try {
            driverClass = classLoader.loadClass(driverClassName).asSubclass(Driver.class);
        }
        catch (ClassNotFoundException | ClassCastException e) {
            throw new RuntimeException("Failed to load Driver class: " + driverClassName, e);
        }

        try {
            return driverClass.getConstructor().newInstance();
        }
        catch (ReflectiveOperationException e) {
            throw new RuntimeException("Failed to create instance of Driver: " + driverClassName, e);
        }
    }

    private static Artifact getArtifact(String driverClass)
    {
        if (driverClass.equals("org.apache.hive.jdbc.HiveDriver")) {
            // Resolve maven artifact at runtime to use different Hive JDBC drivers between Hive and OSS Delta Lake tests because
            // the old Hive JDBC driver (0.13.1) can't connect to new Delta Lake (2.4.0) and
            // the new Hive JDBC driver (>= 3.0.0) supports the above Delta Lake version, but it can't connect to old Hive servers (1.2.1).
            return new DefaultArtifact("org.apache.hive:hive-jdbc:jar:standalone:3.1.3");
        }
        if (driverClass.equals("com.databricks.client.jdbc.Driver")) {
            return new DefaultArtifact("com.databricks:databricks-jdbc:2.6.32");
        }
        throw new IllegalArgumentException("Unexpected driver class: " + driverClass);
    }

    private static JdbcConnectionsPool mockConnectionsPool()
    {
        return new JdbcConnectionsPool()
        {
            @Override
            public Connection connectionFor(JdbcConnectivityParamsState jdbcParamsState)
            {
                throw new UnsupportedOperationException();
            }
        };
    }
}

```

Reply:
@findepi Thanks for sharing the above code. Calling `new DeltaQueryExecutor (...)` per `onDelta()` causes test failure when the test sets session parameter, e.g. `SET spark.databricks.delta.retentionDurationCheck.enabled = false` in `testDatabricksVacuumRemoveChangeDataFeedFiles`. 

Adding another wrapper (`Map<JdbcConnectivityParamsState, JdbcQueryExecutor>`) will resolve the issue though. Do you have any recommendations?

Reply:
i think we need to have these connections managed in per-test context.
This may be easiest done in tempto, but maybe not required. would have to look into how tempto managed context life cycle. maybe we can register stuff in there.

Reply:
@findepi tempto manages the context in `TestInitializationListener`, so injecting something looks a little hard.  
I changed to set a dedicated connection pool for creating new connections. Connection leak won't happen in my understanding because each connection is closed in `JdbcQueryExecutor`. Could you take another look?

Reply:
> tempto manages the context in `TestInitializationListener`, so injecting something looks a little hard.

did you consider registering the connection pool using `io.trino.tempto.context.TestContext#registerCloseCallback`?

Reply:
Yes, I didn't register the pool because a connection is closed by `JdbcConnectionsPool` either. I guess you suggest registering it to avoid holding closed connections there, right? 

Updated to call `registerCloseCallback`.